test/exceed_test.gd

```diff
@@ -0,0 +1,279 @@
+	else:
+		var ws_card_id = player.deck[0].id
+		assert_true(game_logic.do_strike(player, card_id, true, ex_card_id))
+		card_id = ws_card_id
+
+	return card_id
+
+func advance_turn(player):
+	assert_true(game_logic.do_prepare(player))
+	if player.hand.size() > 7:
+		var cards = []
+		var to_discard = player.hand.size() - 7
+		for i in range(to_discard):
+			cards.append(player.hand[i].id)
+		assert_true(game_logic.do_discard_to_max(player, cards))
+
+func validate_gauge(player, amount, id):
+	assert_eq(len(player.gauge), amount)
+	if amount == 0 or len(player.gauge) != amount:
+		return
+	assert_true(
+		player.gauge.any(func (card): return card.id == id),
+		"Didn't find card %s in gauge." % id)
+
+func validate_discard(player, amount, id):
+	assert_eq(len(player.discards), amount)
+	if amount == 0 or len(player.discards) != amount:
+		return
+	assert_true(
+		player.discard.any(func (card): return card.id == id),
+		"Didn't find card %s in discard." % id)
+
+func process_decisions(player, strike_state, decisions):
+	while game_logic.game_state == Enums.GameState.GameState_PlayerDecision and \
+			game_logic.active_strike.strike_state == strike_state and \
+			game_logic.decision_info.player == player.my_id:
+		var content = decisions.pop_front()
+		if content == null:
+			fail_test("Player %s needed to decide on %s during %s but wasn't told how to" % [
+					player.my_id, Enums.DecisionType.keys()[game_logic.decision_info.type],
+					LocalGame.StrikeState.keys()[strike_state]])
+			return
+		match game_logic.decision_info.type:
+			Enums.DecisionType.DecisionType_ForceForEffect:
+				assert_true(game_logic.do_force_for_effect(player, content, false),
+						"%s failed to perform a Force effect using %s" % [player, content])
+			Enums.DecisionType.DecisionType_GaugeForEffect:
+				assert_true(game_logic.do_gauge_for_effect(player, content),
+						"%s failed to perform a Gauge effect using %s" % [player, content])
+			Enums.DecisionType.DecisionType_PayStrikeCost_Required, Enums.DecisionType.DecisionType_PayStrikeCost_CanWild:
+				# There is sometimes an init_extra_cost here
+				# TODO: See if anyone needs it to be more than 0
+				assert_true(game_logic.do_pay_strike_cost(player, content, false),
+						"%s failed to pay a Strike cost using %s" % [player, content])
+			var decision_type:  # Unknown decision type, just roll with it
+				assert_true(game_logic.do_choice(player, content),
+						"Decision of type %s unhandled by test harness (attempted by player %s, content %s)" % [
+								decision_type, player, content])
+
+func process_remaining_decisions(initiator, defender, init_choices, def_choices):
+	var empty_loop_count = 0
+	while init_choices.size() + def_choices.size() >= 1:
+		# empty_loop_count is used to detect when there are still prescribed choices in
+		# the input, but the game is not actually making additional decision points available.
+		empty_loop_count += 1
+		if empty_loop_count >= 3:
```

Comment:
I don't get it, why 3?

Reply:
If a test writer has composed the test incorrectly to specify more decisions than the game actually makes available, the choices.size() check will never fail and will gets stuck in this loop. The `empty_loop_count` check is there to break out of the loop after enough time[1] goes by without being presented with another game decision.

The actual number in this line is just a number of retries. Any small number would be fine, depending on your tolerance for the game advancing in parallel with the test[1].

[1] Some of this is based on the assumption that the game code runs in parallel with the test code, so it is possible for there to be non-negligible clock time between two consecutive decisions if a bunch of other Game Stuff has to happen in between. In those cases we don't want to jump out of the decision-resolving loop too fast. This is also the significance of the TODO (and the commented `wait_seconds`) for the previous comment -- the speed of iteration has to be tuned to some sense of what a tolerable amount of clock time between decisions is.

So then there are two cases to consider when you have leftover decision specs and no decision is being presented: either the test (or the code itself) is incorrectly written, and you should flag a failure; or the game code is just slow, in which case you should just wait a bit for the next decision and give it a chance to pass. The iteration speed and the number of retries in the `empty_loop_count` check reflect the point at which you decide you're in the first case instead of the second.